### PR TITLE
fix(state synchronization): SEG and RT viewports should keep their before hydration voi 

### LINKED ISSUE #3410: [Bug] SEG/RTSTRUCT changes W/L after loading
### Describe the Bug

W/L changes after loading a SEG

### Steps to Reproduce

1. Open a study on OHIFv3 demo: https://v3-demo.ohif.org/viewer?StudyInstanceUIDs=1.3.12.2.1107.5.2.32.35162.30000015050317233592200000046

2. Note the auto W/L=95/35
image

3. Drag the SEG into the viewport

4. Previous W/L has been undone to become W/L=689/315
image

5. Open a study on OHIFv3 demo: https://v3-demo.ohif.org/viewer?StudyInstanceUIDs=1.3.6.1.4.1.14519.5.2.1.3671.4754.298665348758363466150039312520

6. Note the auto W/L in each viewport: A=1048/524; B=3644/1822; C=173/86; D=2400/1300.
image

7. Drag the T2 Weighted Axial SEG into viewport A

8. Previous W/L for viewports A and C have been undone.
image

### The current behavior

W/L changes after loading a SEG

### The expected behavior

W/L should not change after loading a SEG

### OS

Windows 10

### Node version

16.14.0

### Browser

Chrome 111.0.0

### PR DESCRIPTION
**Pull request by sedghi**
fixes #3410 

### Context

As described in the issue above, when having SEG or RT viewport, and the user clicks on the load/hydrate, the old
viewport VOI LUT is not preserved and is removed. The reason for this is that this action is an "update" 
action, and we only storePresentation state on "set" actions (during disable element). 

### Changes & Results

- refactored the storePresentation into a command so that we can call it from both set and update actions.
- In addition the getPresentationIds should not increment the index at the end of the id if the action is an update (
  if in the same viewportIndex and displaySetInstanceUid we are requesting a new presentation id)

### Testing

- https://deploy-preview-3560--ohif-dev.netlify.app/viewer?StudyInstanceUIDs=1.3.6.1.4.1.14519.5.2.1.1706.8374.643249677828306008300337414785
- 1x2 
- d&d SEG on the right
- Invert the left viewport 
- Load the SEG (hydrate)
- see that the left viewport is inverted (was not before the fix)

#### PR

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

- [x] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [] OS: 
- [] Node version: 
- [] Browser:
  

[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch

### CODE REVIEW

**Comment by jbocce** [comment]
I started testing these changes with regards to the steps outlined in #3410. And the way I read that issue, there are TWO problems. The first problem in outlined in steps 1-4 and the second problem in steps 5-8. The second problem appears to be resolved with these changes. However, the first problem appears to remain. Namely...

1. Open this study: https://v3-demo.ohif.org/viewer?StudyInstanceUIDs=1.3.12.2.1107.5.2.32.35162.30000015050317233592200000046
2. Note the auto W/L=95/35
3. Drag the SEG into the viewport
4. Previous W/L has been undone to become W/L=689/315

File: platform/docs/docs/configuration/configurationFiles.md [hunk 1]
```diff
@@ -120,6 +121,7 @@ Here are a list of some options available:
 if auth headers are used, a preflight request is required.
 - `maxNumRequests`: The maximum number of requests to allow in parallel. It is an object with keys of `interaction`, `thumbnail`, and `prefetch`. You can specify a specific number for each type.
 - `showLoadingIndicator`: (default to true), if set to false, the loading indicator will not be shown when navigating between studies.
+- `supportsWildcard`: (default to false), if set to true, the datasource will support wildcard matching for patient name and patient id.
```
Comment (jbocce) [inline platform/docs/docs/configuration/configurationFiles.md:124]: Does it make sense to further specify that this is for study level (QIDO) search?

**Comment by sedghi** [comment]
> started testing these changes with regards to the steps outlined in https://github.com/OHIF/Viewers/issues/3410. And the way I read that issue, there are TWO problems. The first problem in outlined in steps 1-4 and the second problem in steps 5-8. The second problem appears to be resolved with these changes. However, the first problem appears to remain. Namely...

It is not a problem and it is by design. If you change the w/l it will keep it for SEG, if not it will reset it to volume which is intended

File: platform/core/src/services/ViewportGridService/getPresentationIds.ts [hunk 2]
```diff
@@ -109,8 +115,30 @@ const getPresentationIds = (viewport, viewports): PresentationIds => {
     lutPresentationArr.push(uid);
   }
 
-  addUniqueIndex(positionPresentationArr, 'positionPresentationId', viewports);
-  addUniqueIndex(lutPresentationArr, 'lutPresentationId', viewports);
+  // only add unique index if the viewport is getting inserted and not updated
+  const isUpdatingViewport = viewports.find(v => {
```
Comment (jbocce) [inline platform/core/src/services/ViewportGridService/getPresentationIds.ts:119]: The name of this variable suggests that it is a boolean, but instead it is a `Viewport`. So either change its name or use `viewports.some`

File: extensions/cornerstone-dicom-seg/src/viewports/OHIFCornerstoneSEGViewport.tsx [hunk 3]
```diff
@@ -160,6 +168,8 @@ function OHIFCornerstoneSEGViewport(props) {
       viewportIndex,
       segDisplaySet,
     }).then(isHydrated => {
+      storePresentationState();
```
Comment (jbocce) [inline extensions/cornerstone-dicom-seg/src/viewports/OHIFCornerstoneSEGViewport.tsx:171]: Why do we store the presentation after/once the SEG is hydrated? I would think that would be too late with respect to fixing issue #3410, but maybe this is for something else. This comment applies to other similar changes in this PR.

**Comment by jbocce** [comment]
> > started testing these changes with regards to the steps outlined in #3410. And the way I read that issue, there are TWO problems. The first problem in outlined in steps 1-4 and the second problem in steps 5-8. The second problem appears to be resolved with these changes. However, the first problem appears to remain. Namely...
> 
> It is not a problem and it is by design. If you change the w/l it will keep it for SEG, if not it will reset it to volume which is intended

Ok. Thanks. I have verified that it keeps the changed w/l for the SEG.

**Comment by jbocce** [comment]
See my comments.

**Comment by jbocce** [comment]
Hmmm. Is this behaviour expected?


https://github.com/OHIF/Viewers/assets/109477394/ca3f4cba-5c95-4725-94b7-d32a0073590c

**Comment by sedghi** [comment]
Yes it is correct, since you chose the second DS but the SEG is on the third DS.

File: extensions/cornerstone-dicom-seg/src/viewports/OHIFCornerstoneSEGViewport.tsx (hunk 3 above)
Comment (sedghi) [reply extensions/cornerstone-dicom-seg/src/viewports/OHIFCornerstoneSEGViewport.tsx:177]: I added comment, sorry it was not clear you are right

Comment (jbocce) [reply extensions/cornerstone-dicom-seg/src/viewports/OHIFCornerstoneSEGViewport.tsx:177]: Sorry, but I am still a bit confused. The `then` callback is called AFTER the SEG is hydrated, correct? If so, then the first line of the comment you added seems a bit contradictory. But maybe I am missing something.

File: extensions/cornerstone-dicom-rt/src/utils/promptHydrateRT.ts [hunk 4]
```diff
@@ -22,6 +23,12 @@ function promptHydrateRT({
     );
 
     if (promptResult === RESPONSE.HYDRATE_SEG) {
+      if (preHydrateCallbacks?.length) {
```
Comment (jbocce) [inline extensions/cornerstone-dicom-rt/src/utils/promptHydrateRT.ts:26]: Consider simply using `preHydrateCallbacks?.forEach(...)` instead. But no big deal either way.

File: extensions/cornerstone-dicom-seg/src/utils/promptHydrateSEG.ts [hunk 5]
```diff
@@ -17,6 +22,12 @@ function promptHydrateSEG({ servicesManager, segDisplaySet, viewportIndex }) {
     );
 
     if (promptResult === RESPONSE.HYDRATE_SEG) {
+      if (preHydrateCallbacks?.length) {
```
Comment (jbocce) [inline extensions/cornerstone-dicom-seg/src/utils/promptHydrateSEG.ts:25]: Consider simply using `preHydrateCallbacks?.forEach(...)` instead. But no big deal either way.

**Review by jbocce — APPROVED** [review]
Approved.

File: extensions/cornerstone-dicom-seg/src/viewports/OHIFCornerstoneSEGViewport.tsx (hunk 3 above)
Comment (jbocce) [reply extensions/cornerstone-dicom-seg/src/viewports/OHIFCornerstoneSEGViewport.tsx]: Thanks for your changes. However, considering the issue seemed fixed even before the most recent changes, I am not sure if the changes actually contribute to the fix. Just wanted to mention this.

File: extensions/cornerstone/src/Viewport/OHIFCornerstoneViewport.tsx [hunk 6]
```diff
@@ -256,7 +230,9 @@ const OHIFCornerstoneViewport = React.memo(props => {
     setImageScrollBarHeight();
 
     return () => {
-      storePresentation();
+      commandsManager.runCommand('storePresentation', {
+        viewportIndex,
```
Comment (wayfarer3130) [inline extensions/cornerstone/src/Viewport/OHIFCornerstoneViewport.tsx:234]: Please use viewportId throughout - we are trying to remove references to viewportIndex generally.

File: extensions/cornerstone/src/commandsModule.ts [hunk 7]
```diff
@@ -616,6 +617,35 @@ function commandsModule({
       );
       toolGroup.setToolEnabled(ReferenceLinesTool.toolName);
     },
+    storePresentation: ({ viewportIndex }) => {
```
Comment (wayfarer3130) [inline extensions/cornerstone/src/commandsModule.ts:620]: This is much cleaner to have in the commandsModule to allow re-use.

**Review by wayfarer3130 — APPROVED** [review]
Generally good with the changes.  Might change the viewportId, but will leave that up to you to decide.

File: extensions/cornerstone/src/Viewport/OHIFCornerstoneViewport.tsx (hunk 6 above)
Comment (sedghi) [reply extensions/cornerstone/src/Viewport/OHIFCornerstoneViewport.tsx:234]: that is another big PR that I will care later

**Pull request merged**